### PR TITLE
Add pre-commit to do black and other magic

### LINKED ISSUE #3: Add a CI (Github Actions) to run tests across pythons

### PR DESCRIPTION
**Pull request by yarikoptic**
We also added unit-testing (Closes #3)  but I don't know if it would just work -- @feilong should fix that.

### CODE REVIEW

**Comment by feilong** [comment]
@yarikoptic The test failed. Apparently pre-commit wasn't installed properly.

Not sure if the problem originated from the "Cache pre-commit" step:
> Cache not found for input keys: pre-commit-3|/opt/hostedtoolcache/Python/3.11.4/x64|b31884416520dae2a8ac590b473b0ad274d7da7cef1f8c468bb35fd0039f9916

Otherwise I suppose `.[dev]` should include pre-commit?

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+# Cancel active CI runs for a PR before starting another run
+concurrency:
+  group: ${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  pre-commit:
+    name: Pre-commit checks
+    uses: beeware/.github/.github/workflows/pre-commit-run.yml@main
```
Comment (yarikoptic) [inline .github/workflows/ci.yml:16]: I missed that it is not some generic pre-commit action.  I think we should just find/use generic precommit thingie here. Pasting from https://github.com/pre-commit/action

```suggestion
    name: Pre-commit checks
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v3
    - uses: actions/setup-python@v3
    - uses: pre-commit/action@v3.0.0
```

**Comment by yarikoptic** [comment]
here we go -- pre-commit works now, testing doesn't . can fix?

**Comment by feilong** [comment]
> here we go -- pre-commit works now, testing doesn't . can fix?

I guess so. Reminiscent of the joke that after fixing one fail, one gets another 16 fails...

**Comment by feilong** [comment]
Obviously `pytest` was not included as a dependency because I used to only test it locally.
@yarikoptic What's your recommended practice to add it?  Perhaps we can have a "dev" collection of dependencies that includes pytest and all the magics you showed me earlier today?  Currently the section is mostly empty.
```
[project.optional-dependencies]
full = [
    "datalad",
]
```

**Comment by feilong** [comment]
Finally the tests work for Windows as well.
They still fail with Python 3.12-dev, but the error was because 3.12 removed `distutils`, which caused `setuptools` to fail, so it's actually not our package's fault.
Perhaps we can remove support for 3.12 for now, and add it back when it's ready?

**Pull request merged**